File: retos-ts/15_dibujando_tablas.ts
// Resultado 4 ⭐
function drawTable(data: Array<Record<string, string | number>>): string {
    const sizes: number[] = []
    const headers: string[] = []
    for(const v of Object.keys(data[0])){
        const tmp = data.map( ob => `${ob[v]}`)
        const nv: string = v[0].toUpperCase() + v.slice(1)
        tmp.unshift(nv)
        headers.push(nv)
        const max = tmp.reduce((a, b) => 
            a.length > b.length ? a : b, '').length;
        
        sizes.push(max)
    }

    const result: string[] = []
    data.forEach(row => {
        let tmp = '|'
        Object.keys(row).forEach((k,i) => {
            let tstr = `${row[k]}`
            tmp += ` ${tstr.padEnd(sizes[i],' ')} |`
        })
        result.push(tmp)
    })
    
    let sep = '+'
    let fh = '|'
    headers.forEach((h,i) => {
        sep += '-'.repeat(sizes[i] + 2) + '+'
        fh += ` ${h.padEnd(sizes[i],' ')} |`
    })

    result.unshift(sep,fh,sep)
    result.push(sep)

    return result.join('\n')
}

module.exports = drawTable

File: retos-python/15_dibujando_tablas.py
# Resultado 4 ⭐
def draw_table(data):
    sizes = []
    headers = []

    for key in data[0].keys():
        column_values = [str(row[key]) for row in data]
        header = key[0].upper() + key[1:]
        column_values.insert(0, header)
        headers.append(header)
        max_length = max(len(value) for value in column_values)
        sizes.append(max_length)

    result = []

    for row in data:
        row_str = '|'
        for i, key in enumerate(row.keys()):
            value = str(row[key])
            row_str += f" {value.ljust(sizes[i])} |"
        result.append(row_str)

    separator = '+'
    header_row = '|'
    for i, header in enumerate(headers):
        separator += '-' * (sizes[i] + 2) + '+'
        header_row += f" {header.ljust(sizes[i])} |"

    result.insert(0, separator)
    result.insert(1, header_row)
    result.insert(2, separator)
    result.append(separator)

    return '\n'.join(result)
